content/BPIP-3.md

```diff
@@ -146,28 +149,113 @@ The following method is added.
+- true vouchers, with exactly one exchange associated.
+  
+In current protocol, all vouchers are issued when the protocol invokes `issueVoucher` method. In this proposal, we add an additional method `preMint` which allows the seller (operator) to issue a desired number of vouchers.
+Since currently it holds that the exchange id (in the protocol) always matches the token id (in voucher contract), it's desired that this stays even with premint. To achieve it, we propose a method to reserve range (`reserveRange`), which effectively reserves a desired number of exchange ids in the protocol. Then whenever a preminted voucher is converted into a true voucher, it keeps its token id and gets the matching exchange id in the protocol.
+
+Although range reservation ultimately happens on voucher contracts, the method must be invoked true protocol. That way, it can be ensured that the amount of preminted vouchers never exceeds the quantity available, set in the offer.
+
+If the offer expires or is voided, it currently means that buyers cannot commit to it anymore. However, with preminted vouchers, it's possible that some of them were not bought before that happened. In this case, the protocol should also prevent committing to the offer, which means that some unusable vouchers are stuck in the contract and possibly on marketplaces. To delist them, we suggest a method to burn all preminted vouchers (`burnPremintedVouchers`) that cannot be redeemed anymore.
+
+Given that now the buyer does not directly interact with the protocol, this opens the question of how to ensure protocol exchange guarantees (i.e. possibility of getting funds back if something goes wrong). The proposed solution is that now the seller needs to provide enough funds to cover both the item price and the seller deposit. This immediately ensures buyer protection but increases a seller's capital requirement a bit. However, this drawback is not too big, since as soon as the marketplace exchange happens, the seller receives the funds, which can be directly used to cover the item price in subsequent sales. Moreover, as soon as some exchanges are finalized and funds are released back into the seller's pool, those funds can be automatically used to cover future sales.
+
+This proposal might raise some efficiency concerns about the solution efficiency since minting ERC721 is a relatively costly operation. ERC1155 was considered as an alternative, but it turned out that with it it's impossible to maintain a nice exchange id and token id equivalence. To overcome the problem of ERC721 inefficiency, we propose another solution. Since premint always assigns token ownership to the voucher contract owner, there is no need to explicitly store its address during the premint. It's enough that only correct transfer events are emitted, while general consistency is achieved by overriding ERC721 `ownerOf`, `transferFrom` and `safeTransferFrom` methods to correctly manage all tokens that were preminted.
```

Comment:
```suggestion
This proposal might raise some efficiency concerns around the minting of multiple ERC721s, as these are relatively costly operations. ERC1155 was considered as an alternative, but it turned out that with it it's impossible to maintain a nice exchange id and token id equivalence. To overcome the problem of ERC721 inefficiency, we proposed another solution. Since premint always assigns token ownership to the voucher contract owner (or to the contract itself), there is no need to explicitly store its address during the premint phase (saving massively on storage costs). It's enough that only correct transfer events are emitted, while general consistency is achieved by overriding ERC721 `ownerOf`, `transferFrom` and `safeTransferFrom` methods to correctly manage all tokens that were preminted.
```